man/tpm2_encodeobject.1.md
% tpm2_encodeobject(1) tpm2-tools | General Commands Manual

# NAME

**tpm2_encodeobject**(1) - Encode an object into a combined PEM format.

# SYNOPSIS

**tpm2_encodeobject** [*OPTIONS*]

# DESCRIPTION

**tpm2_encodeobject**(1) - Encode both the private and public portions of an
object into a combined PEM format used by tpm2-tss-engine.

The tool reads private and public portions of an object and encodes it
into a combined PEM format used by tpm2-tss-engine and other
applications.

**NOTE**: Both private and public portions of the tpm key must be specified.

# OPTIONS

  * **-C**, **\--parent-context**=_OBJECT_:

    The parent object.

  * **-P**, **\--auth**=_AUTH_:

    The authorization value of the parent object specified by **-C**.

  * **-u**, **\--public**=_FILE_:

    A file containing the public portion of the object.

  * **-r**, **\--private**=_FILE_:

    A file containing the sensitive portion of the object.

  * **-p**, **\--key-auth**:

    Indicates if an authorization value is needed for the object specified by
    **-r** and **-u**.

  * **-o**, **\--output**=_FILE_:

    The output file path, recording the public portion of the object.

## References

[context object format](common/ctxobj.md) details the methods for specifying
_OBJECT_.

[authorization formatting](common/authorizations.md) details the methods for
specifying _AUTH_.

[common options](common/options.md) collection of common options that provide
information many users may expect.

[common tcti options](common/tcti.md) collection of options used to configure
the various known TCTI modules.

# EXAMPLES

## Setup
To load an object you first must create an object under a primary object. So the
first step is to create the primary object.

```bash
tpm2_createprimary -c primary.ctx
```

Step 2 is to create an object under the primary object.

```bash
tpm2_create -C primary.ctx -u key.pub -r key.priv -f pem -o pub.pem
```

This creates the private and public portions of the TPM object. With these
object portions, it is now possible to load that object into the TPM for
subsequent use.

## Encoding an Object into a combined PEM format

The final step, is encoding the public and private portions of the object into a
PEM format.

```bash
tpm2_encodeobject -C primary.ctx -u key.pub -r key.priv -o priv.pem
```

The generated `priv.pem` can be used together with `pub.pem` created in the
step 2 of Setup section.

[returns](common/returns.md)

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/misc/tpm2_encodeobject.c
/* SPDX-License-Identifier: BSD-3-Clause */

/*
 * Part of this file is copied from tpm2-tss-engine.
 *
 * Copyright 2017-2018, Fraunhofer SIT sponsored by Infineon Technologies AG
 * All rights reserved.
 * Copyright (c) 2019, Wind River Systems.
 * All rights reserved.
 */

#include <inttypes.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

#include "files.h"
#include "log.h"
#include "tpm2.h"
#include "tpm2_options.h"
#include "tpm2_tool.h"
#include "object.h"

typedef struct tpm_encodeobject_ctx tpm_encodeobject_ctx;
struct tpm_encodeobject_ctx {
    struct {
        const char *ctx_path;
        const char *auth_str;
        tpm2_loaded_object object;
    } parent;

    struct {
        const char *pubpath;
        TPM2B_PUBLIC public;
        const char *privpath;
        TPM2B_PRIVATE private;
        ESYS_TR handle;
        bool needs_auth;
    } object;

    char *output_path;
};

static tpm_encodeobject_ctx ctx;

static bool on_option(char key, char *value) {
    switch (key) {
    case 'P':
        ctx.parent.auth_str = value;
        break;
    case 'u':
        ctx.object.pubpath = value;
        break;
    case 'r':
        ctx.object.privpath = value;
        break;
    case 'C':
        ctx.parent.ctx_path = value;
        break;
    case 'o':
        ctx.output_path = value;
	break;
    case 'p':
        ctx.object.needs_auth = true;
        break;
    }

    return true;
}

static bool tpm2_tool_onstart(tpm2_options **opts) {
    const struct option topts[] = {
      { "auth",           required_argument, NULL, 'P' },
      { "public",         required_argument, NULL, 'u' },
      { "private",        required_argument, NULL, 'r' },
      { "parent-context", required_argument, NULL, 'C' },
      { "output",         required_argument, NULL, 'o' },
      { "key-auth",       no_argument,       NULL, 'p' },
    };

    *opts = tpm2_options_new("P:u:r:C:o:p", ARRAY_LEN(topts), topts, on_option,
            NULL, 0);

    return *opts != NULL;
}

static tool_rc check_opts(void) {
    tool_rc rc = tool_rc_success;
    if (!ctx.parent.ctx_path) {
        LOG_ERR("Expected parent object via -C");
        rc = tool_rc_option_error;
    }

    if (!ctx.object.pubpath) {
        LOG_ERR("Expected public object portion via -u");
        rc = tool_rc_option_error;
    }

    if (!ctx.object.privpath) {
        LOG_ERR("Expected private object portion via -r");
        rc = tool_rc_option_error;
    }

    if (!ctx.output_path) {
        LOG_ERR("Expected output file path via -o");
        rc = tool_rc_option_error;
    }

    return rc;
}

static tool_rc init(ESYS_CONTEXT *ectx) {
    bool res = files_load_public(ctx.object.pubpath, &ctx.object.public);
    if (!res) {
        return tool_rc_general_error;
    }

    res = files_load_private(ctx.object.privpath, &ctx.object.private);
    if (!res) {
        return tool_rc_general_error;
    }

    return tpm2_util_object_load_auth(ectx, ctx.parent.ctx_path,
            ctx.parent.auth_str, &ctx.parent.object, false,
            TPM2_HANDLE_ALL_W_NV);
}

static int encode(void) {

    uint8_t private_buf[sizeof(ctx.object.private)];
    size_t private_len = 0;
    tool_rc rc = tool_rc_general_error;
    BIO *bio = NULL;
    TSSPRIVKEY_OBJ *tpk = NULL;
    BIGNUM *bn_parent = NULL;
    TSS2_RC rval = Tss2_MU_TPM2B_PRIVATE_Marshal(&ctx.object.private,
        private_buf, sizeof(private_buf), &private_len);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_ERR("Error serializing private portion of object");
        goto error;
    }

    size_t public_len = 0;
    uint8_t public_buf[sizeof(ctx.object.public)];
    rval = Tss2_MU_TPM2B_PUBLIC_Marshal(&ctx.object.public, public_buf,
        sizeof(public_buf), &public_len);
    if (rval != TPM2_RC_SUCCESS) {
        LOG_ERR("Error serializing public portion of object");
        goto error;
    }

    tpk = TSSPRIVKEY_OBJ_new();
    if (!tpk) {
        LOG_ERR("oom");
        goto error;
    }

    tpk->type = OBJ_txt2obj(OID_loadableKey, 1);
    if (!tpk->type) {
        LOG_ERR("oom");
        goto error;
    }

    tpk->emptyAuth = ctx.object.needs_auth;

    bn_parent = BN_new();
    if (!bn_parent) {
        goto error;
    }

    if ((ctx.parent.object.handle >> TPM2_HR_SHIFT) == TPM2_HT_PERSISTENT) {
        BN_set_word(bn_parent, ctx.parent.object.handle);
    } else {
	/* Indicate that the parent is a primary object generated on the fly. */
        BN_set_word(bn_parent, TPM2_RH_OWNER);
    }

    BN_to_ASN1_INTEGER(bn_parent, tpk->parent);
    ASN1_STRING_set(tpk->privkey, private_buf, private_len);
    ASN1_STRING_set(tpk->pubkey, public_buf, public_len);

    bio =  BIO_new_file(ctx.output_path, "w");
    if (bio == NULL) {
	    LOG_ERR("Could not open file: \"%s\"", ctx.output_path);
        goto error;
    }

    PEM_write_bio_TSSPRIVKEY_OBJ(bio, tpk);
    rc = tool_rc_success;

 error:
    BN_free(bn_parent);
    if (bio)
        BIO_free(bio);
    if (tpk)
        TSSPRIVKEY_OBJ_free(tpk);
    return rc;
}

static tool_rc tpm2_tool_onrun(ESYS_CONTEXT *ectx, tpm2_option_flags flags) {
    UNUSED(flags);

    tool_rc rc = check_opts();
    if (rc != tool_rc_success) {
        return rc;
    }

    rc = init(ectx);
    if (rc != tool_rc_success) {
        return rc;
    }

    return encode();
}

// Register this tool with tpm2_tool.c
TPM2_TOOL_REGISTER("encodeobject", tpm2_tool_onstart, tpm2_tool_onrun, NULL, NULL)
